### PR TITLE
Replace hand-rolled utilities with Base and LogExpFunctions equivalents

### PR DESCRIPTION
**Pull request by cossio**
Consolidates internal utility functions onto library or standard-library equivalents, after a survey of every helper in `src/`. All replacements are behavior-preserving (verified against the old formulas, including edge cases).

## Implemented replacements

- **`vstack` / `vwiden` → `Base.stack(...; dims = 1)`** at the gradient/moment sites (`∂cgfs`, `∂energy_from_moments`, `moments_from_samples`), where the stacked arrays are homogeneous broadcast results. The `@declare_layer` keyword constructor instead inlines the equivalent `vcat`-of-reshapes with a comment explaining why: ChainRules has an `rrule` for `stack(::AbstractArray)` but not for tuple arguments (so Zygote fails differentiating through layer conversions like `dReLU(::pReLU)`), and `stack` over a `Vector` loses type inference when the parameters mix container types (a view of `par` alongside a broadcast result), which is common in `shift_fields`, `anneal_zero`, and `_drelu_mixture_moments`. `vcat` satisfies all constraints.
- **`spin_cgf`** now delegates to `LogExpFunctions.logaddexp(θ, -θ)`, whose implementation is the identical `|θ| + log1pexp(-2|θ|)` formula (bit-identical results, checked including `±Inf` and `Float32`).
- **`initialize!(::Binary)`** uses `LogExpFunctions.logit` instead of the hand-written `-log(1/μ - 1)`.
- **`sqrt1half`** uses `Base.hypot`, which handles the large-argument overflow that previously required an eps-based guard branch (`_sqrt1half` deleted; existing edge-case tests for `0`, `±1`, `NaN`, `±Inf`, `1e300`, and `Float32` still pass).
- **`inf` / `two`** type-preserving helpers deleted in favor of `Base.oftype` at their two call sites (`relu_energy`, `tnmean`); value semantics are identical.
- **`collect_states`** uses `Base.stack` instead of a quadratic `reduce`-with-`cat`, keeping the same output shape, eltype, and state order.

## Considered and deliberately not replaced

- `wmean` (StatsBase `mean(A, weights(w))`): would add a dependency and lose the overflow-safe normalization and exact zero-weight annihilation that `test/zero_weight_training.jl` relies on.
- `nobs`/`getobs`/`shuffleobs`/`infinite_minibatches` (MLUtils.jl duplicates these APIs): would add a fairly heavy dependency for ~60 lines; left to a maintainer decision.
- `categorical_rand`/`categorical_sample`, `onehot_encode`/`onehot_decode` (Distributions.jl / OneHotArrays.jl): new dependencies, different data structures.
- `randnt` is a documented, deliberately specialized adaptation of Distributions.jl's sampler (avoids the CDF overhead), kept as-is.
- `_mutable_copy` is exactly `Base.copymutable`, but that name is not public API, so the local helper stays.

## Testing

- Full test suite via `Pkg.test()`.
- Focused runs of `test/layers.jl`, `test/pottsgumbel.jl`, `test/util.jl`, `test/truncnorm.jl`, `test/onehot.jl`, and `test/jlarrays.jl` (JLArrays with `allowscalar(false)`, confirming `Base.stack` is GPU-safe).
- No CHANGELOG entry: internal refactor with no user-facing API or behavior change (none of the removed helpers are in the package's `public` list).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01TnXNNAHv6rrGzMSbE4FbSL

---
_Generated by [Claude Code](https://claude.ai/code/session_01TnXNNAHv6rrGzMSbE4FbSL)_

### CODE REVIEW

File: src/train/initialization.jl [hunk 1]
```diff
@@ -34,7 +34,7 @@ function initialize!(layer::Binary, data::AbstractArray; ϵ::Real = 1.0e-6, wts
     @assert 0 < ϵ < 1 / 2
     μ = batchmean(layer, data; wts)
     μϵ = clamp.(μ, ϵ, 1 - ϵ)
-    @. layer.θ = -log(1 / μϵ - 1)
+    layer.θ .= logit.(μϵ)
```
[the remark this answers is not in the document]
Comment (cossio) [reply src/train/initialization.jl:37]: This is not correct: `layer.θ .= logit.(μϵ)` is a fused broadcast assignment. In Julia, `.=` with a dotted right-hand side lowers to `Base.materialize!(layer.θ, Base.broadcasted(logit, μϵ))`, which writes directly into `layer.θ` with no intermediate array. The previous `@. layer.θ = -log(1 / μϵ - 1)` is macro sugar for exactly the same fused `.=` form, so the two versions have identical allocation behavior. Resolving.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by cossio** [comment]
Closing in favor of focused split PRs, as requested:

- #198 — Replace `vstack`/`vwiden` with `Base.stack`
- #199 — Use LogExpFunctions `logaddexp` and `logit` in spin cgf and Binary initialization
- #200 — Use `Base.hypot` in `sqrt1half`
- #201 — Replace the `inf`/`two` type-preserving helpers with `Base.oftype`
- #202 — Use `Base.stack` in `collect_states`

Together the five PRs reproduce this PR's diff exactly (they are file-disjoint or touch disjoint hunks, each based on current master). CI for the combined diff was fully green here, including both Codecov checks.

---
_Generated by [Claude Code](https://claude.ai/code)_
